libhdate/docs/man/man1/hcal.1
.\" .UC 4
.TH "HCAL" "1" "2 Mar 2005" "Yaacov Zamir" "libhdate"
.SH "NAME"
hcal \- Hebrew date
.PP 
example programs for libhdate, a library that help use hebrew dates.
.PP 
LibHdate is a small C,C++ library for Hebrew dates,
holidays, and reading sequence (parasha). It is using 
the source code from Amos Shapir's "hdate" package fixed and 
patched by Nadav Har'El. The Torah reading sequence
is from tables by Zvi Har'El.
.SH "SYNOPSIS"
.B hcal
[\-hdi]
.PP
[ [ month ] year ]
.SH "DESCRIPTION"
.I Hcal
prints a calendar with both common and Hebrew dates for the
specified month, or the whole year if no month is specified.
If no arguments are given, it prints the current month.
Major holidays are marked by a `*', other holidays by a `+', 
fast days by a `\-'.
.PP 
If the year is greater than 3000,
.I hcal
takes the given date as a Hebrew date, and translates it
back to the common calendar.
Months are numbered with Tishrey being 1 and Elul 12; Adar I and Adar II are months 13 and 14.
.PP
.I \-h 
: Print output in html format.
.PP
.I \-d 
: Use diaspora reading and holidays.
.PP
.I \-i 
: Use external css file "./hcal.css".
.PP
.SH "SEE ALSO"
date(1), cal(1), hdate(1), hebcal(1), remind(1), hdate.h(3), hdate_struct(3)
.SH "AUTHOR"
Yaacov Zamir 2005, text copied from Amos Shapir's hdate man file.


libhdate/ChangeLog
libhdate (1.4.20)

  * Update pascal class example (Patch: Ido Kanner)
  * Update perl and php examples to use hdate objets
  * Fix PHP install dirs, and add hdate.php to the include path
  * Fix hdate man page
  
  -- Yaacov Zamir <kzamir@wall.co.il>  Tue, 5 Nov 2010 10:00

libhdate (1.4.18)

  * Fix longitude positive and negative directions.
  * in the hdate example make the times printout more verbose
  * Add sfirat ha omer to the hdate example (Patch: Amiad Bareli)
  * Add 'bet' in front of the month name in hebrew (Patch: Amiad Bareli)
  
  -- Yaacov Zamir <kzamir@wall.co.il>  Tue, 26 Oct 2010 22:00
  
libhdate (1.4.14)

  * Fix --localedir configure option and add a gettext localedir indicator
  	at the configure summary
  * Add Sfirat ha omer function (Patch: Moshe Wagner <moshe.wagner@gmail.com>)
  * Add Erev Yum Kipur to the holidays list 
  	(Bug report: Arnon Sharlin; Patch: Moshe Wagner <moshe.wagner@gmail.com>)
  * Fix Tora readings in diaspora (Patch: Moshe Wagner <moshe.wagner@gmail.com>)
  * Change the type of holiday function to me more descriptive
  	(Patch: Moshe Wagner <moshe.wagner@gmail.com>)
  	
  -- Yaacov Zamir <kzamir@wall.co.il>  Saturday, 17 Oct 2008 22:00
  
libhdate (1.4.12)

  * Add GNU Pascal bindings (Ido Kanner)
  * Add example for GNU pascal
  * change configure flags for pascal from --disable-pascal
      to --disable-fpc and --disable-gpc

  -- Yaacov Zamir <kzamir@wall.co.il>  Saturday, 14 Jun 2008 22:00
  
libhdate (1.4.11)

  * Add Ruby bindings
  * Add PHP bindings
  * Add GPL notice to the example files (Bug report: Baruch Even)
  * Add a patch to print todays date in hcal in bold (Patch: Patrik Nilsson)
  * Fix documentation of files and structs (Bug report: Baruch Even)
  * Update Swedish translation (Thanks: Patrik Nilsson)
  * Fix hdate_get_utc_sun_time_deg 
      function only work in low altitudes (Bug report: Ido Kanner)
  * Update pascal bindings and examples (Patch: Ido Kanner)

  -- Yaacov Zamir <kzamir@wall.co.il>  Saturday, 5 Jan 2008 22:00
  
libhdate (1.4.10)

  * Move to GPL3 (COPYING, AUTHORS)
      - Add Amos Shapir as author in Authors file
      - Update Amos's email
  * Change defult config options (configure.in)
      - Gettext version used now is 0.17
      - Python, Perl, Pascal bindings now built by default
  * Update the man files (docs/man)
      - add man files for c++ language
      - move all man files to docs/man directory
  * Bug fixes
      - Fix Kislev Hebrew translation misspell (Bug report: Ariel Galola)
      - Fix Julian day calculation (Bug report: Afonso Arantes)
  * Add new events
      - 30 shvat : Family day
      - 12 heshvan : rabin memorial day
      - 29 tamuz : zhabotinsky day
  * Changes to hdate command line utility
      - flag -j print the Julian day number
      - if argument is greater than 100000 it is asumed as Julian day number
      - "hdate -S YEAR" now print out same format as "hebcal -d -h -e YEAR"
  
  -- Yaacov Zamir <kzamir@wall.co.il>  Thusday, 3 January 2008 22:00

libhdate (1.4.9)

  * Fix endianess bug in big endian machins (/examples/hcal/hcal.c and hdate.c)
	
  -- Yaacov Zamir <kzamir@wall.co.il>  Saturday, 7 April 2007 22:00

libhdate (1.4.8)

  * Fix help printing bug: Add new line at the end of hdate help text.
  * Add two new functios to the time of day section:
      hdate_get_utc_sun_time_full - gets the time of first light,
      talit, sunris, sunset, first stars and three stars.
      hdate_get_utc_sun_time_deg - gets the time the sun is on
      some known altitude.
  * Add Times of day to hdate command line tool.
     supported times: first-light, talit, sunrise, mid-day, sunset, first-stars
     and three-stars
  * Add Times of Shabat enter/leave to hdate command line tool.
     Shabat starts 20 min before sunset, and leave at three stars.
  * Add new oprions in hdate command line tool to it's man page
	
  -- Yaacov Zamir <kzamir@wall.co.il>  Tue, 8 Jun 2006 12:13
  
libhdate (1.4.7)

  * Fix pesach II in hebrew po file.
  * hcal utility export strict html 4.01.
  * hcal utility can export diaspora holidays
  * Update hcal utility man file.

  -- Yaacov Zamir <kzamir@wall.co.il>  Tue, 18 Apr 2006 12:13

libhdate (1.4.3)

  * Fix header of spanish po file.
  * Daniel Nylander: Swedish translation.

  -- Yaacov Zamir <kzamir@wall.co.il>  Mon, 10 Apr 2006 12:13
  
libhdate (1.4.2)

  * Ido Kanner <idokan@gmail.com> - Update free pascal bindings.
  * Update address of Free Software Foundation in COPYING file.
  * make the spec.in file more configurable.

  -- Yaacov Zamir <kzamir@wall.co.il>  Fri, 25 Nov 2005 12:13
  
libhdate (1.4.0)

  * Michael Kaminsky - libhdate now reports the 2nd day of yom
    tov on Sukkot and Peseach as Yom Tov and not Chol Hamoed.
  * Cange C++ API - insert the location and diaspora flag as
    part of the hdate object.
  * Docomentation - separate the c and the cpp docs, more readable.
  * hdate example can now export iCal formated output.

  -- Yaacov Zamir <kzamir@wall.co.il>  Mon, 8 Aug 2005 22:13
  
libhdate (1.3.1)

  * Add "const correctness" patch from Markus Elfring (elfring).
  * Manual page seems to contain a hyphen where a minus sign was 
intended, patch by Lior Kaplan.
  * Remove the debian directory from the ziped package.

    -- Yaacov Zamir <kzamir@wall.co.il>  Sat, 30 Jul 2005 22:13

libhdate (1.3.0)

  * Add examples to the doc directory
  * doc directory is now /usr/doc/libhdate0 and no /usr/doc/hdate
  * make int_to_string aware of LANGUAGE system variable, make it more gettext compatible

    -- Yaacov Zamir <kzamir@wall.co.il>  Wed, 27 Jul 2005 18:13

libhdate (1.0.1)

  * first major reales number.
  * start to number the shared library.
  * stable api.
  * updated utility programs hdate/hcal.
  * added man files for library and utility programs.
  * python and perl bindings.
  * free pascal binding by Ido Kanner <idokan@gmail.com>
  * ru, fr, he, es translations, thanks to Peled Esti and Boris Peterberg.

libhdate/examples/hcal/hcal.c

/* hcal.c
 * Example file for libhdate. 
 * 
 * compile:
 * gcc `pkg-config --libs --cflags libhdate` hcal.c -o hcal
 * 
 * Copyright:  2004 (c) Yaacov Zamir 
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>		/* For printf */
#include <hdate.h>		/* For hebrew date */
#include <stdlib.h>		/* For atoi */
#include <locale.h>		/* For setlocale */
#include <unistd.h>		/* For getopt */

#define FALSE 0
#define TRUE -1

/* print help */
int
print_help (char *program)
{
	printf ("hcal - example program for libhdate\n\n");

	printf ("USAGE: %s [-hid] ", program);
	printf ("   [[month] year]\n");

	printf ("OPTIONS:\n");
	printf ("   -h : Print html format.\n");
	printf ("   -d : Use diaspora reading and holidays.\n");
	printf ("   -i : Use external css file \"./hcal.css\".\n");

	return 0;
}

/* print html css part */
int
print_css (int opt_i)
{
	if (opt_i)
	{
		printf ("\n\t@import \"hcal.css\";\n");

	}
	else
	{
		printf ("\n\
body {\n\
  direction: rtl;\n\
}\n\
\n\
img { \n\
    margin:0;\n\
    padding: 0;\n\
    vertical-align: middle;\n\
    border: 0;\n\
}\n\
\n\
p {\n\
  \n\
}\n\
\n\
table {\n\
	width: 80%;\n\
	table-layout: fixed;\n\
	font-size: 14pt;\n\
	border: solid #aaaaaa;\n\
}\n\
\n\
th {\n\
	background-color: #aaaaaa;\n\
	text-align: center;\n\
}\n\
\n\
td {\n\
	text-align: center;\n\
}\n\
\n\
div.today {\n\
\n\
}\n\
\n\
div.gmonth {\n\
	font-size: 16pt;\n\
}\n\
\n\
div.gyear {\n\
	font-size: 24pt;\n\
	font-weight: bold;\n\
}\n\
\n\
div.gday {\n\
	\n\
}\n\
\n\
div.hmonth {\n\
	font-size: 16pt;\n\
}\n\
\n\
div.hyear {\n\
	font-size: 24pt;\n\
	font-weight: bold;\n\
}\n\
\n\
div.hday {\n\
	\n\
}\n\
\n\
div.holiday_name {\n\
	\n\
}\n\
\n\
td.sat {\n\
	border: solid #777777;\n\
}\n\
\n\
td.regular {\n\
	border: solid #aaaaaa;\n\
}\n\
\n\
td.holiday {\n\
	color: #990000;\n\
	border: solid #888888;\n\
}\n\
\n\
td.out_of_month {\n\
	color: #dddddd;\n\
	border: solid #dddddd;\n\
}\n");
	}

	return 0;
}

/* print html header */
int
print_html_header (int opt_i)
{
	printf ("\
<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\"\n\
   \"http://www.w3.org/TR/html4/strict.dtd\">\n\
<html>\n\
<head>\n\
<meta name=\"generator\" content=\"hcal (libhdate)\">\n\
<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\">\n\
<style title=\"Normal\" type=\"text/css\" media=\"all\">");

	print_css (opt_i);

	printf ("</style>\n");

	/* some alternative css styles */
	printf ("<link rel=\"alternate stylesheet\" title=\"High contrast\" \
type=\"text/css\" media=\"screen\" href=\"high_contrast.css\">");
	printf ("</style>\n\
<link rel=\"alternate stylesheet\" title=\"Colorful\" \
type=\"text/css\" media=\"screen\" href=\"colorful.css\">");
	printf ("</style>\n\
<link rel=\"alternate stylesheet\" title=\"Print\" \
type=\"text/css\" media=\"all\" href=\"print.css\">\n");

	printf ("\
<title>Hebrew Calendar</title>\n\
</head>\n\
<body>\n");

	return 0;
}

int
print_html_footer ()
{
	printf ("<!-- <p>\n\
<a href=\"http://validator.w3.org/check/referer\">\n<img \
src=\"http://www.w3.org/Icons/valid-html401\"\n\
alt=\"Valid HTML 4.01!\" height=\"31\" width=\"88\" />\n</a>\n");
	printf ("<a href=\"http://jigsaw.w3.org/css-validator\">\n\
<img src=\"http://www.w3.org/Icons/valid-css\" alt=\"Valid CSS!\">\n\
</a>\n</p> -->\n");

	printf ("</body>\n</html>\n");

	return 0;
}

/* print year and month part */
int
print_header (int month, int year, int opt_h, int opt_d)
{
	hdate_struct h1, h2;
	int j;

	/* set dates for begining and end of calendar */
	hdate_set_gdate (&h1, 1, month, year);
	hdate_set_gdate (&h2, 1, month % 12 + 1, year);

	/* Print Gregorian month and year */
	if (opt_h)
	{
		printf ("<div class=\"gmonth\">\n");
		printf ("%s\n", hdate_get_month_string (h1.gd_mon, FALSE));
		printf ("</div>\n");

		printf ("<div class=\"gyear\">\n");
		printf ("%d\n", h1.gd_year);
		printf ("</div>\n");
	}
	else
	{
		printf ("%s %d\n", hdate_get_month_string (h1.gd_mon, FALSE),
			h1.gd_year);
	}

	/* Print Hebrew month and year */
	if (opt_h)
	{
		printf ("<div class=\"hmonth\">\n");
	}

	if (h1.hd_mon != h2.hd_mon)
	{
		printf ("%s-",
			hdate_get_hebrew_month_string (h1.hd_mon, FALSE));
	}
	printf ("%s ", hdate_get_hebrew_month_string (h2.hd_mon, FALSE));

	if (opt_h)
	{
		printf ("\n</div>\n");

		printf ("<div class=\"hyear\">\n");
	}

	printf ("%s\n", hdate_get_int_string (h1.hd_year));

	if (opt_h)
	{
		printf ("</div>\n");

		printf ("<div class=\"month_table\">\n");
		printf ("<table>\n");
		printf ("<tr>\n");
	}

	for (j = 1; j < 8; j++)
	{
		if (opt_h)
		{
			printf ("<th>");
			printf ("%3s", hdate_get_day_string (j, FALSE));
			printf ("</th>\n");
		}
		else
		{
			printf ("%3s", hdate_get_day_string (j, TRUE));

			if (j != 7)
				printf ("\t");
		}
	}

	if (opt_h)
	{
		printf ("</tr>\n");
	}
	else
	{
		printf ("\n");
	}

	return 0;
}

/* print month table */
int
print_calendar (int month, int year, int opt_h, int opt_d)
{
	hdate_struct h;
	hdate_struct today;
	int jd;
	int jd_today;
	int i, j;
	char type_char[] = { '/', '+', '*', '-' };
	int holyday_type;

	/* Find day to start calendar with */
	hdate_set_gdate (&h, 1, month, year);

	/* Set today to today, so we know what day to highlight */
	hdate_set_gdate (&today, 0, 0, 0);
	jd_today = today.hd_jd;
	
	/* Return print head to sunday */
	jd = h.hd_jd - h.hd_dw + 1;

	/* Loop over all days in this month */
	for (i = 0; i < 6; i++)
	{
		if (opt_h)
		{
			printf ("<tr>\n");
		}

		for (j = 0; j < 7; j++)
		{
			/* Get this day hebrew date */
			hdate_set_jd (&h, jd);

			/* Get this day holyday type ba harez (diaspora flag =
			 * 0) */
			holyday_type =
				hdate_get_holyday_type (hdate_get_holyday
							(&h, opt_d));

			if (opt_h)
			{
				if (h.gd_mon != month)
				{
					printf ("<td class=\"out_of_month\">\n");
				}
				else if (j == 6)
				{
					printf ("<td class=\"sat\">\n");
				}
				else if (holyday_type)
				{
					printf ("<td class=\"holiday\">\n");
				}
				else
				{
					printf ("<td class=\"regular\">\n");
				}

				/* Print a day */
				printf ("<div class=\"gday\">%2d</div>\n<div class=\"hday\">%3s</div>\n", h.gd_day, hdate_get_int_string (h.hd_day));

				if (holyday_type)
				{
					printf ("<div class=\"holiday_name\">%s</div>\n", hdate_get_holyday_string (hdate_get_holyday (&h, opt_d), FALSE));
				}

				printf ("</td>\n");
			}
			else
			{
				if (h.gd_mon == month)
				{
					/* Print a day */
					if (jd != jd_today)
					{
						printf ("%2d%c%3s", h.gd_day,
							type_char[holyday_type],
							hdate_get_int_string (h.
									      hd_day));
					}
					else /* It's today, lets print it in bold */
					{
						printf ("%c[1m", 27);
						printf ("%2d%c%3s", h.gd_day,
							type_char[holyday_type],
							hdate_get_int_string (h.
									      hd_day));
						printf ("%c[m", 27);
					}
				}
				if (j != 6)
					printf ("\t");
			}
			jd++;
		}

		if (opt_h)
		{
			printf ("</tr>\n");
		}
		else
		{
			printf ("\n");
		}
	}

	if (opt_h)
	{
		printf ("</table>\n</div>");
	}

	return 0;
}

/* print month header and month table */
int
print_month (int month, int year, int opt_h, int opt_d)
{
	hdate_struct h;

	/* check if hebrew year (year > 3000) */
	if (year > 3000)
		hdate_set_hdate (&h, 1, month, year);
	else
		hdate_set_gdate (&h, 1, month, year);

	/* Print calendar header */
	print_header (h.gd_mon, h.gd_year, opt_h, opt_d);
	print_calendar (h.gd_mon, h.gd_year, opt_h, opt_d);

	printf ("\n");

	return 0;
}

int
main (int argc, char *argv[])
{
	/* date */
	int month;
	int year;

	/* user opts */
	int c;
	int opt_h = 0;		/* -h html format flag */
	int opt_d = 0;		/* -d Diaspora holidays */
	int opt_i = 0;		/* -i External css file */

	/* hdate struct */
	hdate_struct h;

	/* init locale */
	setlocale (LC_ALL, "");

	/* command line parsing */
	while ((c = getopt (argc, argv, "shdi")) != -1)
	{
		switch (c)
		{
		case 'h':
			opt_h = 1;
			break;
		case 'd':
			opt_d = 1;
			break;
		case 'i':
			opt_i = 1;
			break;
		default:
			print_help (argv[0]);
			exit (0);
			break;
		}
	}

	/* Get date from user */
	if (argc == (optind + 2))
	{
		month = atoi (argv[optind]);
		year = atoi (argv[optind + 1]);
	}
	else if (argc == (optind + 1))
	{
		month = 0;
		year = atoi (argv[optind]);

		/* Check for valid years */
		if (year < 1)
		{
			/* Print help for user and exit */
			print_help (argv[0]);
			exit (0);
		}
	}
	else if (argc == (optind))
	{
		hdate_set_gdate (&h, 0, 0, 0);	/* get today's year */
		month = h.gd_mon;
		year = h.gd_year;
	}
	else
	{
		/* Print help for user and exit */
		print_help (argv[0]);
		exit (0);
	}

	/* if html print html header */
	if (opt_h)
		print_html_header (opt_i);

	/* print all year */
	if (month == 0)
	{
		for (month = 1; month < 13; month++)
		{
			print_month (month, year, opt_h, opt_d);
		}
	}
	else			/* print only this month */
	{
		print_month (month, year, opt_h, opt_d);
	}

	/* if html print html header */
	if (opt_h)
		print_html_footer ();

	return 0;
}
